### PR TITLE
Refactor pounders

### PR DESCRIPTION
**Pull request by janosg**
Refactor pounders algorithm to prepare for later extensions.

### CODE REVIEW

File: tests/optimization/test_pounders_integration.py [hunk 1]
```diff
@@ -40,12 +42,12 @@ def func(x: np.ndarray, exog: np.ndarray, endog: np.ndarray) -> np.ndarray:
 
 start_params = [
     np.array([0.15, 0.008, 0.01]),
-    np.ones(3) * 0.25,
+    # np.ones(3) * 0.25, # noqa: E800
```
Comment (janosg) [inline tests/optimization/test_pounders_integration.py:45]:
```suggestion
```

**Review by janosg — COMMENTED** [review]
Thanks! Just a few minor comments. 

Can you please also replace all the pickle files with test fixtures by something human readable like `json`, `csv` or `yaml` (depending on what information it is)?

Since we both worked on this PR we'll also need a separate approval by @timmens.

File: src/estimagic/optimization/history.py [hunk 2]
```diff
@@ -0,0 +1,247 @@
+"""History class for pounders and similar optimizers."""
+import numpy as np
+
+
+class LeastSquaresHistory:
+    """Container to save and retrieve history entries for a least-squares optimizer."""
```
Comment (timmens) [inline src/estimagic/optimization/history.py:6]: Looking at this class alone its not entirely clear what a "history" is. Maybe add one or two sentences to the docstring.

**Review by timmens — APPROVED** [review]

File: tests/optimization/test_pounders_integration.py (hunk 1 above)
Comment (segsell) [reply tests/optimization/test_pounders_integration.py]: I deleted the line ```# np.ones(3) * 0.25, # noqa: E800``` or do you want to keep it for future reference?

**Pull request merged**